components/cluster/command/deploy.go

```diff
@@ -94,360 +61,46 @@ func newDeploy() *cobra.Command {
-				iterErr = errors.New(
-					"'imported' is set to 'true' for new instance, this is only used " +
-						"for instances imported from tidb-ansible and make no sense when " +
-						"deploying new instances, please delete the line or set it to 'false' for new instances")
-				return // skip the host to avoid issues
+			topoFile := args[2]
+			if data, err := ioutil.ReadFile(topoFile); err == nil {
+				teleTopology = string(data)
 			}
 
-			uniqueHosts[inst.GetHost()] = hostInfo{
-				ssh:  inst.GetSSHPort(),
-				os:   inst.OS(),
-				arch: inst.Arch(),
-			}
-			var dirs []string
-			for _, dir := range []string{globalOptions.DeployDir, globalOptions.LogDir} {
-				if dir == "" {
-					continue
-				}
-				dirs = append(dirs, clusterutil.Abs(globalOptions.User, dir))
-			}
-			// the default, relative path of data dir is under deploy dir
-			if strings.HasPrefix(globalOptions.DataDir, "/") {
-				dirs = append(dirs, globalOptions.DataDir)
-			}
-			t := task.NewBuilder().
-				RootSSH(
-					inst.GetHost(),
-					inst.GetSSHPort(),
-					opt.user,
-					sshConnProps.Password,
-					sshConnProps.IdentityFile,
-					sshConnProps.IdentityFilePassphrase,
-					gOpt.SSHTimeout,
-				).
-				EnvInit(inst.GetHost(), globalOptions.User).
-				Mkdir(globalOptions.User, inst.GetHost(), dirs...).
-				BuildAsStep(fmt.Sprintf("  - Prepare %s:%d", inst.GetHost(), inst.GetSSHPort()))
-			envInitTasks = append(envInitTasks, t)
-		}
-	})
-
-	if iterErr != nil {
-		return iterErr
-	}
-
-	// Download missing component
-	downloadCompTasks = prepare.BuildDownloadCompTasks(clusterVersion, &topo)
-
-	// Deploy components to remote
-	topo.IterInstance(func(inst spec.Instance) {
-		version := spec.ComponentVersion(inst.ComponentName(), clusterVersion)
-		deployDir := clusterutil.Abs(globalOptions.User, inst.DeployDir())
-		// data dir would be empty for components which don't need it
-		dataDirs := clusterutil.MultiDirAbs(globalOptions.User, inst.DataDir())
-		// log dir will always be with values, but might not used by the component
-		logDir := clusterutil.Abs(globalOptions.User, inst.LogDir())
-		// Deploy component
-		// prepare deployment server
-		t := task.NewBuilder().
-			UserSSH(inst.GetHost(), inst.GetSSHPort(), globalOptions.User, gOpt.SSHTimeout).
-			Mkdir(globalOptions.User, inst.GetHost(),
-				deployDir, logDir,
-				filepath.Join(deployDir, "bin"),
-				filepath.Join(deployDir, "conf"),
-				filepath.Join(deployDir, "scripts")).
-			Mkdir(globalOptions.User, inst.GetHost(), dataDirs...)
-
-		// copy dependency component if needed
-		switch inst.ComponentName() {
-		case spec.ComponentTiSpark:
-			t = t.DeploySpark(inst, version, "" /* default srcPath */, deployDir)
-		default:
-			t = t.CopyComponent(
-				inst.ComponentName(),
-				inst.OS(),
-				inst.Arch(),
+			return deployer.Deploy(
+				clusterName,
 				version,
-				"", // use default srcPath
-				inst.GetHost(),
-				deployDir,
+				topoFile,
+				opt,
+				afterDeployForDeployCluster,
```

Comment:
It's better use "post" instead of "after", and as this is an unexported function, I think it can be with some simpler name like `postDeployActions` or `postDeployHooks` or similar thing.